# Project 7: Winning Jeopardy

# Objective:


Jeopardy is a popular TV show in the US where participants answer questions to win money. It's been running for a few decades, and is a major force in popular culture. Let's say we want to compete on Jeopardy, and we are looking for any edge that we can get to win. 

In this project, we will work with a dataset of Jeopardy questions to figure out some patterns in the questions that could help us to win. The dataset is freely available on the link (https://www.reddit.com/r/datasets/comments/1uyd0t/200000_jeopardy_questions_in_a_json_file/) as JEOPARDY_CSV.csv file.

# Introduction to Dataset:

The dataset contains 216930 rows of Jeopardy questions. Each row in the dataset represents a single question on a single episode of Jeopardy. Here are explanations of each column:

Show Number -- the Jeopardy episode number of the show this question was in.
Air Date -- the date the episode aired.
Round -- the round of Jeopardy that the question was asked in. Jeopardy has several rounds as each episode progresses.
Category -- the category of the question.
Value -- the number of dollars answering the question correctly is worth.
Question -- the text of the question.
Answer -- the text of the answer. 

# Reading in Dataset:

In [2]:
import pandas as pd
import csv
jeopardy = pd.read_csv('JEOPARDY_CSV.csv', sep=',')
jeopardy.head(10)

Show Number    Air Date      Round                         Category  Value  \
0         4680  2004-12-31  Jeopardy!                          HISTORY   $200   
1         4680  2004-12-31  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES   $200   
2         4680  2004-12-31  Jeopardy!      EVERYBODY TALKS ABOUT IT...   $200   
3         4680  2004-12-31  Jeopardy!                 THE COMPANY LINE   $200   
4         4680  2004-12-31  Jeopardy!              EPITAPHS & TRIBUTES   $200   
5         4680  2004-12-31  Jeopardy!                   3-LETTER WORDS   $200   
6         4680  2004-12-31  Jeopardy!                          HISTORY   $400   
7         4680  2004-12-31  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES   $400   
8         4680  2004-12-31  Jeopardy!      EVERYBODY TALKS ABOUT IT...   $400   
9         4680  2004-12-31  Jeopardy!                 THE COMPANY LINE   $400   

                                            Question          Answer  
0  For the last 8 years of his life, Galileo was ...      Copernicus  
1  No. 2: 1912 Olympian; football star at Carlisl...      Jim Thorpe  
2  The city of Yuma in this state has a record av...         Arizona  
3  In 1963, live on "The Art Linkletter Show", th...      McDonald's  
4  Signer of the Dec. of Indep., framer of the Co...      John Adams  
5  In the title of an Aesop fable, this insect sh...         the ant  
6  Built in 312 B.C. to link Rome & the South of ...  the Appian Way  
7  No. 8: 30 steals for the Birmingham Barons; 2,...  Michael Jordan  
8  In the winter of 1971-72, a record 1,122 inche...      Washington  
9  This housewares store was named for the packag...  Crate & Barrel

In [3]:
jeopardy.columns

Index(['Show Number', ' Air Date', ' Round', ' Category', ' Value',
       ' Question', ' Answer'],
      dtype='object')

#Removing white spaces from column names.

In [4]:
jeopardy.columns = ['Show Number', 'Air Date', 'Round', 'Category', 'Value', 'Question', 'Answer']

# Normalizing String in columns:

Before we can start doing analysis on the Jeopardy questions, we need to normalize all of the text columns (the Question and Answer columns), this is because of lowercase words and punctuations. For example Don't and don't aren't considered to be different words when we compare them.

In [5]:
import re

def normalize_text(text):
    text = text.lower()
    text = re.sub("[^A-Za-z0-9\s]", "", text)
    return text

In [6]:
jeopardy["clean_question"] = jeopardy["Question"].apply(normalize_text)
jeopardy["clean_answer"] = jeopardy["Answer"].apply(normalize_text)
jeopardy.head(10)

Show Number    Air Date      Round                         Category Value  \
0         4680  2004-12-31  Jeopardy!                          HISTORY  $200   
1         4680  2004-12-31  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  $200   
2         4680  2004-12-31  Jeopardy!      EVERYBODY TALKS ABOUT IT...  $200   
3         4680  2004-12-31  Jeopardy!                 THE COMPANY LINE  $200   
4         4680  2004-12-31  Jeopardy!              EPITAPHS & TRIBUTES  $200   
5         4680  2004-12-31  Jeopardy!                   3-LETTER WORDS  $200   
6         4680  2004-12-31  Jeopardy!                          HISTORY  $400   
7         4680  2004-12-31  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  $400   
8         4680  2004-12-31  Jeopardy!      EVERYBODY TALKS ABOUT IT...  $400   
9         4680  2004-12-31  Jeopardy!                 THE COMPANY LINE  $400   

                                            Question          Answer  \
0  For the last 8 years of his life, Galileo was ...      Copernicus   
1  No. 2: 1912 Olympian; football star at Carlisl...      Jim Thorpe   
2  The city of Yuma in this state has a record av...         Arizona   
3  In 1963, live on "The Art Linkletter Show", th...      McDonald's   
4  Signer of the Dec. of Indep., framer of the Co...      John Adams   
5  In the title of an Aesop fable, this insect sh...         the ant   
6  Built in 312 B.C. to link Rome & the South of ...  the Appian Way   
7  No. 8: 30 steals for the Birmingham Barons; 2,...  Michael Jordan   
8  In the winter of 1971-72, a record 1,122 inche...      Washington   
9  This housewares store was named for the packag...  Crate & Barrel   

                                      clean_question    clean_answer  
0  for the last 8 years of his life galileo was u...      copernicus  
1  no 2 1912 olympian football star at carlisle i...      jim thorpe  
2  the city of yuma in this state has a record av...         arizona  
3  in 1963 live on the art linkletter show this c...       mcdonalds  
4  signer of the dec of indep framer of the const...      john adams  
5  in the title of an aesop fable this insect sha...         the ant  
6  built in 312 bc to link rome  the south of ita...  the appian way  
7  no 8 30 steals for the birmingham barons 2306 ...  michael jordan  
8  in the winter of 197172 a record 1122 inches o...      washington  
9  this housewares store was named for the packag...   crate  barrel

Now that we've normalized the text columns, there are also some other columns to normalize. The Value column should also be numeric, in order to manipulate it more easily. We'll need to remove the dollar sign from the beginning of each value and convert the column from text to numeric. 

In [7]:
def normalize_values(text):
    text = re.sub("[^A-Za-z0-9\s]", "", text)
    try:
        text = int(text)
    except Exception:
        text = 0
    return text

In [8]:
jeopardy["clean_value"] = jeopardy["Value"].apply(normalize_values)

In [9]:
jeopardy.head(10)

Show Number    Air Date      Round                         Category Value  \
0         4680  2004-12-31  Jeopardy!                          HISTORY  $200   
1         4680  2004-12-31  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  $200   
2         4680  2004-12-31  Jeopardy!      EVERYBODY TALKS ABOUT IT...  $200   
3         4680  2004-12-31  Jeopardy!                 THE COMPANY LINE  $200   
4         4680  2004-12-31  Jeopardy!              EPITAPHS & TRIBUTES  $200   
5         4680  2004-12-31  Jeopardy!                   3-LETTER WORDS  $200   
6         4680  2004-12-31  Jeopardy!                          HISTORY  $400   
7         4680  2004-12-31  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  $400   
8         4680  2004-12-31  Jeopardy!      EVERYBODY TALKS ABOUT IT...  $400   
9         4680  2004-12-31  Jeopardy!                 THE COMPANY LINE  $400   

                                            Question          Answer  \
0  For the last 8 years of his life, Galileo was ...      Copernicus   
1  No. 2: 1912 Olympian; football star at Carlisl...      Jim Thorpe   
2  The city of Yuma in this state has a record av...         Arizona   
3  In 1963, live on "The Art Linkletter Show", th...      McDonald's   
4  Signer of the Dec. of Indep., framer of the Co...      John Adams   
5  In the title of an Aesop fable, this insect sh...         the ant   
6  Built in 312 B.C. to link Rome & the South of ...  the Appian Way   
7  No. 8: 30 steals for the Birmingham Barons; 2,...  Michael Jordan   
8  In the winter of 1971-72, a record 1,122 inche...      Washington   
9  This housewares store was named for the packag...  Crate & Barrel   

                                      clean_question    clean_answer  \
0  for the last 8 years of his life galileo was u...      copernicus   
1  no 2 1912 olympian football star at carlisle i...      jim thorpe   
2  the city of yuma in this state has a record av...         arizona   
3  in 1963 live on the art linkletter show this c...       mcdonalds   
4  signer of the dec of indep framer of the const...      john adams   
5  in the title of an aesop fable this insect sha...         the ant   
6  built in 312 bc to link rome  the south of ita...  the appian way   
7  no 8 30 steals for the birmingham barons 2306 ...  michael jordan   
8  in the winter of 197172 a record 1122 inches o...      washington   
9  this housewares store was named for the packag...   crate  barrel   

   clean_value  
0          200  
1          200  
2          200  
3          200  
4          200  
5          200  
6          400  
7          400  
8          400  
9          400

Similarly the Air Date column should also be a datetime, not a string, to enable us to work with it more easily.

In [10]:
jeopardy["Air Date"] = pd.to_datetime(jeopardy["Air Date"])

In [11]:
jeopardy.dtypes

Show Number                int64
Air Date          datetime64[ns]
Round                     object
Category                  object
Value                     object
Question                  object
Answer                    object
clean_question            object
clean_answer              object
clean_value                int64
dtype: object

# Questions under Study:

In order to figure out whether to study past questions, study general knowledge, or not study it all, it would be helpful to figure out two things:

Q1:How often the answer is deducible from the question?
Q2:How often new questions are repeats of older questions?

In order to answer first question we will write a function to seeing how many times words in the answer also occur in the question. To answer the second question we will see how often complex words (> 6 characters) reoccur in the question column.

# Answer terms in the question

In [12]:
def count_matches(row):
    split_answer = row["clean_answer"].split(" ")
    split_question = row["clean_question"].split(" ")
    if "the" in split_answer:
        split_answer.remove("the")
    if len(split_answer) == 0:
        return 0
    match_count = 0
    for item in split_answer:
        if item in split_question:
            match_count += 1
    return match_count / len(split_answer)

jeopardy["answer_in_question"] = jeopardy.apply(count_matches, axis=1)
mean_percentage = (jeopardy["answer_in_question"].mean())*100
mean_percentage

5.935758718396862

The answer only appears in the question about 5.93% of the time. This isn't a huge number, and means that we probably can't just hope that hearing a question will enable us to figure out the answer. We'll probably have to study.

# Question overlap:

In [13]:
question_overlap = []
terms_used = set()
for i, row in jeopardy.iterrows():
        split_question = row["clean_question"].split(" ")
        split_question = [q for q in split_question if len(q) > 5]
        match_count = 0
        for word in split_question:
            if word in terms_used:
                match_count += 1
        for word in split_question:
            terms_used.add(word)
        if len(split_question) > 0:
            match_count /= len(split_question)
        question_overlap.append(match_count)
jeopardy["question_overlap"] = question_overlap
mean_percentage = jeopardy["question_overlap"].mean()*100

In [14]:
mean_percentage

87.35125558086732

There is about 87.35% overlap between terms in new questions and terms in old questions. This only looks at a small set of questions, and it doesn't look at phrases, it looks at single terms. This makes it relatively insignificant, but it does mean that it's worth looking more into the recycling of questions.

# Low value vs high value questions

Let's say we only want to study questions that pertain to high value questions instead of low value questions. This will help us earn more money when we're on Jeopardy.

We can actually figure out which terms correspond to high-value questions using a chi-squared test. We'll first need to narrow down the questions into two categories:

Low value -- Any row where Value is less than 800.
High value -- Any row where Value is greater than 800.

In [15]:
def determine_value(row):
    value = 0
    if row["clean_value"] > 800:
        value = 1
    return value

jeopardy["high_value"] = jeopardy.apply(determine_value, axis=1)

In [16]:
def count_usage(term):
    low_count = 0
    high_count = 0
    for i, row in jeopardy.iterrows():
        if term in row["clean_question"].split(" "):
            if row["high_value"] == 1:
                high_count += 1
            else:
                low_count += 1
    return high_count, low_count

comparison_terms = list(terms_used)[:5]
observed_expected = []
for term in comparison_terms:
    observed_expected.append(count_usage(term))

observed_expected

[(0, 1), (5, 9), (0, 1), (20, 55), (5, 10)]

# Chi-squared Test

In [17]:
from scipy.stats import chisquare
import numpy as np

high_value_count = jeopardy[jeopardy["high_value"] == 1].shape[0]
low_value_count = jeopardy[jeopardy["high_value"] == 0].shape[0]

chi_squared = []
for obs in observed_expected:
    total = sum(obs)
    total_prop = total / jeopardy.shape[0]
    high_value_exp = total_prop * high_value_count
    low_value_exp = total_prop * low_value_count
    
    observed = np.array([obs[0], obs[1]])
    expected = np.array([high_value_exp, low_value_exp])
    chi_squared.append(chisquare(observed, expected))

chi_squared

[Power_divergenceResult(statistic=0.39497646423335131, pvalue=0.52969509124866954),
 Power_divergenceResult(statistic=0.3777145872319056, pvalue=0.5388288955084517),
 Power_divergenceResult(statistic=0.39497646423335131, pvalue=0.52969509124866954),
 Power_divergenceResult(statistic=0.10029832265881709, pvalue=0.75147192850840205),
 Power_divergenceResult(statistic=0.18617046944535684, pvalue=0.66612320033728945)]

None of the terms had a significant difference in usage between high value and low value rows. Additionally, the frequencies were all lower than 5, so the chi-squared test isn't as valid. It would be better to run this test with only terms that have higher frequencies.